samza-core/src/main/java/org/apache/samza/zk/ZkBarrier.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.zk;
+
+import org.I0Itec.zkclient.IZkChildListener;
+import org.I0Itec.zkclient.IZkDataListener;
+import org.apache.commons.collections4.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+
+
+/**
+ * ZkBarrier is an implementation of distributed barrier, which guarantees that the expected barrier size and barrier
+ * participants match before marking the barrier as complete. It also allows the caller to expire the barrier.
+ *
+ * This implementation is specifically tailored towards barrier support during jobmodel version upgrades. The participant
+ * responsible for the version upgrade starts the barrier by invoking {@link #create(String, List)}.
+ * Each participant in the list, then, joins the new barrier. When all listed participants {@link #join(String, String)}
+ * the barrier, the creator marks the barrier as done which signals the end of barrier.
+ * The creator of the barrier can invoke {@link #expire(String)} to indicate that the barrier has timed_out or
+ * is no longer valid.
+ *
+ * The caller can listen to events associated with the barrier by registering a {@link ZkBarrierListener}.
+ *
+ * Zk Tree Reference:
+ * /barrierRoot/
+ *  |
+ *  |- barrier_{version1}/
+ *  |   |- barrier_state/
+ *  |   |  ([DONE|TIMED_OUT])
+ *  |   |- barrier_participants/
+ *  |   |   |- {id1}
+ *  |   |   |- {id2}
+ *  |   |   |-  ...
+ */
+public class ZkBarrier {
+  private final static Logger LOG = LoggerFactory.getLogger(ZkBarrier.class);
+  private static final String BARRIER_PARTICIPANTS = "/barrier_participants";
+  private static final String BARRIER_STATE = "/barrier_state";
+  private final ZkUtils zkUtils;
+  private final BarrierKeyBuilder keyBuilder;
+
+  public enum State {
+    TIMED_OUT, DONE
+  }
+
+  private ZkBarrierListener barrierListener;
+
+  public ZkBarrier(String barrierRoot, ZkUtils zkUtils, ZkBarrierListener barrierListener) {
```

Comment:
Just curious why this isn't final? do we expect this to change outside the constructor? 
If it is nullable, can we make it `Optional` instead? that way, inside your code you can get rid of null checks and do `barrierListener.ifPresent(...)` It would alleviate concerns over null checks on every code path and force defensive programming inherently. 
Thoughts?